### ISSUE TITLE
SSH agent provider error on Apple M1 Monterey 

### ISSUE THREAD

**Issue by kodeninja** [earthly/earthly#1476]
Getting the following error when running `earthly` on the `hello-world+hello` example:

```bash
$ earthly github.com/earthly/hello-world+hello

 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
Error: ssh agent provider: stat /private/tmp/com.apple.launchd.WpS1Yi0ZWC/Listeners: no such file or directory
```

Environment:
 - Arch: `Apple M1`
 - OS: Monterey
 - Docker Desktop: `4.3.0 (71786)`

```bash
$ docker version
Client: Docker Engine - Community
 Version:           20.10.11
 API version:       1.41
 Go version:        go1.17.2
 Git commit:        dea9396e18
 Built:             Wed Nov 17 23:49:46 2021
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.11
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.9
  Git commit:       847da18
  Built:            Thu Nov 18 00:34:44 2021
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by kodeninja** [comment]
Another data point (not sure if it's relevant though)--when I first ran the above, I had `Use the new Virtualization framework` _checked_ under Docker Desktop's **Experimental Features**. However, I turned it off, restarted DD, re-ran the above command, but still got the same error 🤷🏼‍♂️.

**Comment by camerondurham** [comment]
I unfortunately don't have much to add besides a working data point. Was curious to try this out on new M1 machine and using almost exactly same versions as mentioned and the hello-world example works.

- Arch: `arm64`, Apple M1, 2020
- OS: Monterrey 12.0.1 (also 12.1, just upgraded today)
- Docker Desktop 4.3.1 (72247)

```
earthly github.com/earthly/hello-world+hello

 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)


 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

             context | --> git context github.com/earthly/hello-world+hello
      busybox:1.32.0 | --> Load metadata linux/arm64
g/e/hello-world+base | --> FROM busybox:1.32.0
g/e/hello-world+base | [██████████] 100% resolve docker.io/library/busybox:1.32.0@sha256:bde48e1751173b709090c2539fdf12d6ba64e88ec7a4301591227ce925f3c678
g/e/hello-world+hello | *cached* --> COPY globe.txt ./
g/e/hello-world+hello | *cached* --> RUN cat globe.txt
              output | --> exporting outputs
...

========================== 🌍 Earthly Build  ✅ SUCCESS ==========================
```


```
docker version
Client:
 Cloud integration: v1.0.22
 Version:           20.10.11
 API version:       1.41
 Go version:        go1.16.10
 Git commit:        dea9396
 Built:             Thu Nov 18 00:36:09 2021
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.11
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.9
  Git commit:       847da18
  Built:            Thu Nov 18 00:34:44 2021
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Hopefully you've found a solution but wanted to mention it appears to work fine. I'm on a relatively new install of Monterey and have done very few modifications besides github ssh setup, docker install and almost no Rosetta emulated programs.

**Comment by blackeyeddan** [comment]
I'm also ran into this problem today and was able to work around it by running `ssh-agent` just before running earthly.

Environment:

- Arch: Apple M1 Max
- OS: Monterey 12.1
- Docker Desktop: 4.3.2 (72729)

**Comment by vladaionescu** [comment]
@adamgordonbell reports that this is not M1-specific.

**Comment by alexcb** [comment]
We're having issues consistently reproducing this, but I put together a version of earthly that will try a direct `os.Stat` and will printf results or errors if anyone wants to try running it:

# step 1: compile the new binary:

```
mkdir ~/earthly-testing
cd ~/earthly-testing
SSH_AUTH_SOCK="" earthly github.com/earthly/earthly:a013ec1a4be31f8c407fc2464148d257c5cff2cc+for-darwin
```

(note that the `SSH_AUTH_SOCK=""` will temporarily disable ssh entirely, which should work-around this issue. Notice that we're explicitly not exporting this, as we want to be able to replicate it in the next step).

If you're curious in the code changes, here they are: https://github.com/earthly/earthly/commit/a013ec1a4be31f8c407fc2464148d257c5cff2cc

# step 2: run the new earthly binary
```
./build/darwin/<arch>/earthly github.com/earthly/hello-world+hello
```

This should printout some lines such as:
```
ACB calling os.Stat directly with "/private/tmp/com.apple.launchd.QJMlOfsZvV/Listeners"
ACB os.Stat returned &{Listeners 0 16777654 {735733586 63777794583 0x26810c0} {16777221 49590 1 59471185 501 0 0 [0 0 0 0] {1642197783 735733586} {1642197783 735733586} {1642197783 735810986} {1642197783 735733586} 0 0 4096 0 0 0 [0 0]}}
ACB calling NewSSHAgentProvider with "/private/tmp/com.apple.launchd.QJMlOfsZvV/Listeners"
```

possibly (hopefully?) containing errors that we can use to debug this issue.

# step 3: optional debugging steps outside of earthly (assuming `SSH_AUTH_SOCK` is set):
 - can you `ls $SSH_AUTH_SOCK`?
 - can you `stat $SSH_AUTH_SOCK`?

**Comment by adamgordonbell** [comment]
I have a reproduction, although it's possible this is not the cause of the issue others are having:
 Building something with earthly I get:
```
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Starting buildkit daemon as a docker container (earthly-buildkitd)...
           buildkitd | ...Done
Error: ssh agent provider: stat /private/tmp/com.apple.launchd.dvl5KJemx3/Listeners: no such file or directory
```

If I open a new terminal and build it the debug earthly and run it everything works:
```
./build/darwin/amd64/earthly github.com/earthly/hello-world+hello

 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
           buildkitd | Updated image available. Restarting buildkit daemon...
           buildkitd | ...Done


 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

             context | --> git context github.com/earthly/hello-world+hello
      busybox:1.32.0 | --> Load metadata linux/amd64
g/e/hello-world+base | --> FROM busybox:1.32.0
g/e/hello-world+base | [██████████] 100% resolve docker.io/library/busybox:1.32.0@sha256:bde48e1751173b709090c2539fdf12d6ba64e88ec7a4301591227ce925f3c678
g/e/hello-world+hello | *cached* --> COPY globe.txt ./
g/e/hello-world+hello | *cached* --> RUN cat globe.txt
              output | --> exporting outputs


 3. Push ⏫ (disabled)
————————————————————————————————————————————————————————————————————————————————

To enable pushing use

                earthly --push ...



 4. Local Output 🎁
————————————————————————————————————————————————————————————————————————————————



========================== 🌍 Earthly Build  ✅ SUCCESS ==========================
```
e 
However, it seems in my new terminal this works:
```
ls "$SSH_AUTH_SOCK" 
/private/tmp/com.apple.launchd.LULm3TPcjg/Listeners
```
Whereas in my old one, where earthly failed, it is somehow set wrong:
```
ls "$SSH_AUTH_SOCK"
ls: /private/tmp/com.apple.launchd.dvl5KJemx3/Listeners: No such file or directory
```

This is on a X86 mbp and using the VS Code terminal.

**Comment by camerondurham** [comment]
@adamgordonbell 

No difference for me between iTerm2 and VSCode terminal
```
ls $SSH_AUTH_SOCK 
/private/tmp/com.apple.launchd.maw86CwLbE/Listeners
```

**Comment by alexcb** [comment]
Any chance we can get some more detail about who owns `$SSH_AUTH_SOCK`?

Can you try running:

```
sudo whoami; echo $SSH_AUTH_SOCK; sudo lsof | grep "$SSH_AUTH_SOCK"; ls -la $SSH_AUTH_SOCK; earthly github.com/earthly/hello-world+hello; ls -la $SSH_AUTH_SOCK; sudo lsof | grep "$SSH_AUTH_SOCK"
```

**Comment by alexcb** [comment]
Another user in slack got a similar error: `Error: ssh agent provider: stat /tmp/ssh-CznuCpabULXU/agent.3237: no such file or directory`.

It's possible the user accidentally set `SSH_AUTH_SOCK` without realising it.

as for the MacOS issue -- it could very well be some special permissions which is preventing children of the terminal to inherit permission to access the `SSH_AUTH_SOCK` file.

**Comment by alexcb** [comment]
tl;dr: if you run into this issue, and don't need ssh access, you can explicitly set the variable to be empty before running earthly:

```
SSH_AUTH_SOCK="" earthly +my-target
```
